src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.Core.cs

```diff
@@ -50,7 +51,14 @@ private object DeserializeObject(int typeIndex)
 
             if (_binaryFormatter == null)
             {
-                InitializeBinaryFormatter();
+                if (!InitializeBinaryFormatter())
```

Comment:
I'm not seeing this code pattern handled in an optimal way by the linker. Here's the resulting trimmed code:

```C#
        private object DeserializeObject(int typeIndex)
        {
            if (!this._permitDeserialization)
            {
                throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization);
            }
            if (this._binaryFormatter == null)
            {
                this.InitializeBinaryFormatter();
                throw new NotSupportedException(SR.BinaryFormatter_SerializationDisallowed);
            }
            Type type = this.FindType(typeIndex);
            object sDeserializeMethod = ResourceReader.s_deserializeMethod(this._binaryFormatter, this._store.BaseStream);
            if (sDeserializeMethod.GetType() != type)
            {
                throw new BadImageFormatException(SR.Format(SR.BadImageFormat_ResType_SerBlobMismatch, type.FullName, sDeserializeMethod.GetType().FullName));
            }
            return sDeserializeMethod;
        }
```

Since the switch is cut off inside of the `if (_binaryFormatter == null)`, the linker won't trim the other code in this method. Being able to trim the rest of the code in this method is nice because other things can be trimmed as well, like `s_deserializeMethod`. It might not be a huge deal here since the dead code is minimal. But in general we should try to make the linker check optimal so the rest of the method can be trimmed.

Reply:
@eerhardt Would something like this work? I can send it as a separate PR.

```cs
object binaryFormatter = GetBinaryFormatter(); // hardcoded to return null on wasm
if (binaryFormatter is null)
{
    throw new PNSE();
}

Type type = this.FindType(typeIndex);
// remainder of method

private object BinaryFormatter()
{
    if (_binaryFormatter is null)
    {
        // initialize
    }

    Debug.Assert(_binaryFormatter != null);
    return _binaryFormatter;
}
```

Reply:
I do not think the linker can stub methods that return `object`: https://github.com/mono/linker/blob/811c69dfb6e325a344b713890408b137b39a888b/src/linker/Linker.Steps/BodySubstituterStep.cs#L172

I think the easiest way to fix this is to move `if (this._binaryFormatter == null)` inside `InitializeBinaryFormatter`. Also, I think it would be fine to defer this to https://github.com/dotnet/runtime/issues/32862 . The `InitializeBinaryFormatter` method will need to be rewritten to make it work well with reflection linking and analysis.

Reply:
Note that #32862 is not planned for 5.0.

I am also fine with deferring this right now. I don't see a lot of code being preserved here since the `FindType` method is also called elsewhere. So the size savings won't be impactful.